src/middleware/redirectIfNeedsAuthn.js

```diff
@@ -32,7 +32,11 @@ module.exports = safeAsync(async (req, res, next) => {
   const jwtCookie = req.cookies.jwt
   const user = await getUserFromJwt(jwtCookie)
   if (user === null) {
-    res.redirect(withBaseUrl('/login'))
+    let url = withBaseUrl('/login')
+    if (path !== withBaseUrl('/')) {
```

Comment:
I think we should compare with `withBaseUrl('')` as well - hitting https://edudev.cs.illinois.edu/q with no trailing slash is a valid way to access the index, but it'll end up with `?redirect=/` with the current code